File: lib/domain/entities/game-object.ts
import { Transform } from "./transform"
import { nanoid } from "nanoid"

export interface GameObjectData {
  id?: string
  name: string
  transform?: Transform
  tags?: string[]
  children?: GameObjectData[]
}

export class GameObject {
  id: string
  name: string
  transform: Transform
  tags: string[]
  children: GameObject[]
  parent: GameObject | null = null

  constructor(data: GameObjectData) {
    this.id = data.id || nanoid()
    this.name = data.name
    this.transform = data.transform || new Transform()
    this.tags = data.tags || []
    this.children = (data.children || []).map((child) => new GameObject(child))

    // Set parent references
    this.children.forEach((child) => {
      child.parent = this
    })
  }

  addChild(child: GameObject) {
    this.children.push(child)
    child.parent = this
  }

  removeChild(child: GameObject) {
    const index = this.children.indexOf(child)
    if (index > -1) {
      this.children.splice(index, 1)
      child.parent = null
    }
  }

  findById(id: string): GameObject | null {
    if (this.id === id) return this

    for (const child of this.children) {
      const found = child.findById(id)
      if (found) return found
    }

    return null
  }

  hasTag(tag: string): boolean {
    return this.tags.includes(tag)
  }

  addTag(tag: string) {
    if (!this.tags.includes(tag)) {
      this.tags.push(tag)
    }
  }

  removeTag(tag: string) {
    const index = this.tags.indexOf(tag)
    if (index > -1) {
      this.tags.splice(index, 1)
    }
  }

  toJSON(): GameObjectData {
    return {
      id: this.id,
      name: this.name,
      transform: this.transform.toJSON(),
      tags: [...this.tags],
      children: this.children.map((child) => child.toJSON()),
    }
  }

  static fromJSON(data: GameObjectData): GameObject {
    return new GameObject({
      ...data,
      transform: data.transform ? Transform.fromJSON(data.transform) : new Transform(),
      children: data.children?.map((child) => GameObject.fromJSON(child)),
    })
  }
}

File: lib/domain/scene/game_object.dart
import 'package:mobx/mobx.dart';
import 'game_component.dart';

class GameObject {
  final String id;
  String name;
  String? parentId;

  /// A única fonte de verdade: lista de componentes
  final List<GameComponent> components;

  /// Observable children list so MobX observers detect add/remove
  final ObservableList<GameObject> children;

  GameObject({
    required this.id,
    required this.name,
    this.parentId,
    Map<String, String>? tags,
    List<GameComponent>? components,
    List<GameObject>? children,
  })  : components = components ?? [],
        children = ObservableList.of(children ?? []) {
  }

  // --- Component helpers ---
  T? getComponent<T extends GameComponent>() {
    for (final c in components) {
      if (c is T) return c;
    }
    return null;
  }

  GameComponent? getComponentById(String id) {
    try {
      return components.firstWhere((c) => c.id == id);
    } catch (_) {
      return null;
    }
  }

  void setComponent<T extends GameComponent>(T component) {
    components.removeWhere((c) => c.id == component.id);
    components.add(component);
  }

  /// Return a new GameObject with the given component replaced/added (immutable helper)
  GameObject copyWithComponent(GameComponent newComponent) {
    final newComponents = components.where((c) => c.id != newComponent.id).toList();
    newComponents.add(newComponent);

    return GameObject(
      id: id,
      name: name,
      parentId: parentId,
      components: newComponents,
      children: children,
    );
  }

  void addChild(GameObject child) {
    child.parentId = id;
    children.add(child);
  }

  void removeChild(String childId) {
    children.removeWhere((c) => c.id == childId);
  }

  // --- Manual JSON (polymorphic components) ---
  factory GameObject.fromJson(Map<String, dynamic> json) {
    final id = json['id'] as String;
    final name = json['name'] as String;
    final parentId = json['parentId'] as String?;

    final List<GameComponent> comps = [];
    if (json['components'] != null) {
      final compMap = json['components'] as Map<String, dynamic>;
      compMap.forEach((typeId, data) {
        try {
          final c = ComponentRegistry.create(typeId, data as Map<String, dynamic>);
          comps.add(c);
        } catch (e) {
          // ignore unknown component types
        }
      });
    }

    final kids = <GameObject>[];
    if (json['children'] != null) {
      final list = json['children'] as List;
      kids.addAll(list.map((c) => GameObject.fromJson(c as Map<String, dynamic>)));
    }

    return GameObject(id: id, name: name, parentId: parentId, components: comps, children: kids);
  }

  Map<String, dynamic> toJson() {
    final compsMap = <String, dynamic>{};
    for (final c in components) {
      compsMap[c.id] = c.toJson();
    }

    return {
      'id': id,
      'name': name,
      'parentId': parentId,
      'components': compsMap,
      'children': children.map((c) => c.toJson()).toList(),
    };
  }
}
